File: TarsCli/Services/Workflow/IWorkflowCoordinator.cs
namespace TarsCli.Services.Workflow;

/// <summary>
/// Interface for workflow coordinators
/// </summary>
public interface IWorkflowCoordinator
{
    /// <summary>
    /// Starts a workflow
    /// </summary>
    /// <param name="workflowType">Type of workflow to start</param>
    /// <param name="parameters">Workflow parameters</param>
    /// <returns>Workflow instance</returns>
    Task<WorkflowInstance> StartWorkflowAsync(string workflowType, Dictionary<string, object> parameters);

    /// <summary>
    /// Gets the status of a workflow
    /// </summary>
    /// <param name="workflowId">ID of the workflow</param>
    /// <returns>Workflow instance</returns>
    Task<WorkflowInstance> GetWorkflowStatusAsync(string workflowId);

    /// <summary>
    /// Stops a workflow
    /// </summary>
    /// <param name="workflowId">ID of the workflow</param>
    /// <returns>True if the workflow was stopped successfully, false otherwise</returns>
    Task<bool> StopWorkflowAsync(string workflowId);

    /// <summary>
    /// Gets all active workflows
    /// </summary>
    /// <returns>List of active workflow instances</returns>
    Task<List<WorkflowInstance>> GetActiveWorkflowsAsync();

    /// <summary>
    /// Gets all workflows of a specific type
    /// </summary>
    /// <param name="workflowType">Type of workflow</param>
    /// <returns>List of workflow instances</returns>
    Task<List<WorkflowInstance>> GetWorkflowsByTypeAsync(string workflowType);

    /// <summary>
    /// Transitions a workflow to the next state
    /// </summary>
    /// <param name="workflowId">ID of the workflow</param>
    /// <param name="nextState">Next state to transition to</param>
    /// <param name="result">Result of the current state</param>
    /// <returns>Updated workflow instance</returns>
    Task<WorkflowInstance> TransitionWorkflowAsync(string workflowId, string nextState, object result);
}

/// <summary>
/// Represents a workflow instance
/// </summary>
public class WorkflowInstance
{
    /// <summary>
    /// ID of the workflow
    /// </summary>
    public string Id { get; set; }

    /// <summary>
    /// Type of the workflow
    /// </summary>
    public string Type { get; set; }

    /// <summary>
    /// Current state of the workflow
    /// </summary>
    public string CurrentState { get; set; }

    /// <summary>
    /// Status of the workflow
    /// </summary>
    public WorkflowStatus Status { get; set; }

    /// <summary>
    /// Parameters of the workflow
    /// </summary>
    public Dictionary<string, object> Parameters { get; set; } = new();

    /// <summary>
    /// Results of the workflow
    /// </summary>
    public Dictionary<string, object> Results { get; set; } = new();

    /// <summary>
    /// History of the workflow
    /// </summary>
    public List<WorkflowStateTransition> History { get; set; } = new();

    /// <summary>
    /// Error message if the workflow failed
    /// </summary>
    public string ErrorMessage { get; set; }

    /// <summary>
    /// Start time of the workflow
    /// </summary>
    public System.DateTime StartTime { get; set; }

    /// <summary>
    /// End time of the workflow
    /// </summary>
    public System.DateTime? EndTime { get; set; }
}

/// <summary>
/// Represents a workflow state transition
/// </summary>
public class WorkflowStateTransition
{
    /// <summary>
    /// Previous state
    /// </summary>
    public string FromState { get; set; }

    /// <summary>
    /// Next state
    /// </summary>
    public string ToState { get; set; }

    /// <summary>
    /// Timestamp of the transition
    /// </summary>
    public System.DateTime Timestamp { get; set; }

    /// <summary>
    /// Result of the state
    /// </summary>
    public object Result { get; set; }
}

/// <summary>
/// Status of a workflow
/// </summary>
public enum WorkflowStatus
{
    /// <summary>
    /// Workflow is created but not started
    /// </summary>
    Created,

    /// <summary>
    /// Workflow is running
    /// </summary>
    Running,

    /// <summary>
    /// Workflow is paused
    /// </summary>
    Paused,

    /// <summary>
    /// Workflow is completed successfully
    /// </summary>
    Completed,

    /// <summary>
    /// Workflow failed
    /// </summary>
    Failed,

    /// <summary>
    /// Workflow was cancelled
    /// </summary>
    Cancelled
}

File: TarsEngine.FSharp.Core/CodeGen/Workflow/IWorkflowCoordinator.fs
namespace TarsEngine.FSharp.Core.CodeGen.Workflow

open System
open System.Threading
open System.Threading.Tasks

/// <summary>
/// Represents the status of a workflow.
/// </summary>
type WorkflowStatus =
    | NotStarted
    | Running
    | Completed
    | Failed
    | Cancelled

/// <summary>
/// Represents a workflow step.
/// </summary>
type WorkflowStep = {
    /// <summary>
    /// The name of the step.
    /// </summary>
    Name: string
    
    /// <summary>
    /// The description of the step.
    /// </summary>
    Description: string
    
    /// <summary>
    /// The action to execute for the step.
    /// </summary>
    Action: unit -> Task<obj>
    
    /// <summary>
    /// The status of the step.
    /// </summary>
    mutable Status: WorkflowStatus
    
    /// <summary>
    /// The result of the step, if any.
    /// </summary>
    mutable Result: obj option
    
    /// <summary>
    /// The error message, if any.
    /// </summary>
    mutable ErrorMessage: string option
}

/// <summary>
/// Represents a workflow.
/// </summary>
type Workflow = {
    /// <summary>
    /// The ID of the workflow.
    /// </summary>
    Id: Guid
    
    /// <summary>
    /// The name of the workflow.
    /// </summary>
    Name: string
    
    /// <summary>
    /// The description of the workflow.
    /// </summary>
    Description: string
    
    /// <summary>
    /// The steps in the workflow.
    /// </summary>
    Steps: WorkflowStep list
    
    /// <summary>
    /// The status of the workflow.
    /// </summary>
    mutable Status: WorkflowStatus
    
    /// <summary>
    /// The start time of the workflow, if any.
    /// </summary>
    mutable StartTime: DateTime option
    
    /// <summary>
    /// The end time of the workflow, if any.
    /// </summary>
    mutable EndTime: DateTime option
    
    /// <summary>
    /// Additional information about the workflow.
    /// </summary>
    AdditionalInfo: Map<string, string>
}

/// <summary>
/// Interface for coordinating workflows.
/// </summary>
type IWorkflowCoordinator =
    /// <summary>
    /// Creates a workflow.
    /// </summary>
    /// <param name="name">The name of the workflow.</param>
    /// <param name="description">The description of the workflow.</param>
    /// <param name="steps">The steps in the workflow.</param>
    /// <returns>The created workflow.</returns>
    abstract member CreateWorkflow : name:string * description:string * steps:(string * string * (unit -> Task<obj>)) list -> Workflow
    
    /// <summary>
    /// Executes a workflow.
    /// </summary>
    /// <param name="workflow">The workflow to execute.</param>
    /// <param name="cancellationToken">Optional cancellation token.</param>
    /// <returns>The executed workflow.</returns>
    abstract member ExecuteWorkflowAsync : workflow:Workflow * ?cancellationToken:CancellationToken -> Task<Workflow>
    
    /// <summary>
    /// Gets a workflow by ID.
    /// </summary>
    /// <param name="id">The ID of the workflow.</param>
    /// <returns>The workflow, if found.</returns>
    abstract member GetWorkflow : id:Guid -> Workflow option
    
    /// <summary>
    /// Gets all workflows.
    /// </summary>
    /// <returns>The list of all workflows.</returns>
    abstract member GetAllWorkflows : unit -> Workflow list
    
    /// <summary>
    /// Gets the status of a workflow.
    /// </summary>
    /// <param name="id">The ID of the workflow.</param>
    /// <returns>The status of the workflow, if found.</returns>
    abstract member GetWorkflowStatus : id:Guid -> WorkflowStatus option
    
    /// <summary>
    /// Cancels a workflow.
    /// </summary>
    /// <param name="id">The ID of the workflow.</param>
    /// <returns>Whether the workflow was cancelled.</returns>
    abstract member CancelWorkflow : id:Guid -> bool
